src/kernel/binary_reduce_impl.h

```diff
@@ -0,0 +1,414 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file kernel/binary_reduce_impl.h
+ * \brief Implementations of binary reduce operations.
+ */
+#ifndef DGL_KERNEL_BINARY_REDUCE_IMPL_H_
+#define DGL_KERNEL_BINARY_REDUCE_IMPL_H_
+
+#include <minigun/minigun.h>
+#include <dgl/runtime/device_api.h>
+#include <dgl/immutable_graph.h>
+
+#include <algorithm>
+#include <string>
+
+#ifdef __CUDACC__
+#include "../runtime/cuda/cuda_common.h"
+#endif
+#include "./binary_reduce.h"
+#include "./binary_reduce_impl_decl.h"
+#include "./utils.h"
+
+namespace dgl {
+namespace kernel {
+
+/****************************************************
+ * BinaryOpReduce
+ ****************************************************/
+template <int XPU, typename Idx, typename DType, typename Reducer>
+GData<Idx, DType> AllocGData(
+    const DLContext& ctx, int64_t x_len,
+    runtime::NDArray lhs_mapping, runtime::NDArray rhs_mapping,
+    runtime::NDArray lhs_data, runtime::NDArray rhs_data,
+    runtime::NDArray out_mapping, runtime::NDArray out_data) {
+  // GData
+  GData<Idx, DType> gdata;
+  gdata.x_length = x_len;
+  gdata.out_size = out_data->shape[0];
+  gdata.lhs_data = static_cast<DType*>(lhs_data->data);
+  gdata.rhs_data = static_cast<DType*>(rhs_data->data);
+  gdata.out_data = static_cast<DType*>(out_data->data);
+  if (!utils::IsNoneArray(lhs_mapping)) {
+    gdata.lhs_mapping = static_cast<Idx*>(lhs_mapping->data);
+  }
+  if (!utils::IsNoneArray(rhs_mapping)) {
+    gdata.rhs_mapping = static_cast<Idx*>(rhs_mapping->data);
+  }
+  if (!utils::IsNoneArray(out_mapping)) {
+    gdata.out_mapping = static_cast<Idx*>(out_mapping->data);
+  }
+  // fill out data with zero values
+  utils::Fill<XPU>(ctx, gdata.out_data, utils::NElements(out_data), Zero<Reducer>::value);
+  return gdata;
```

Comment:
what is GData?

Reply:
More comments (in binary_reduce_impl_decl.h). Pls checkout.